src/canisters/frontend/ic-asset/src/upload.rs

```diff
@@ -42,20 +47,22 @@ pub async fn upload(
     )
     .await?;
 
-    let operations = assemble_batch_operations(
-        project_assets,
-        canister_assets,
-        AssetDeletionReason::Incompatible,
-    );
-
-    info!(logger, "Committing batch.");
-
-    let args = CommitBatchArguments {
-        batch_id,
-        operations,
-    };
-
-    commit_batch(canister, args).await?;
+    match api_version(canister).await {
+        0 => {
+            let operations = batch_upload::operations::v0::assemble_batch_operations(
```

Comment:
I recommend against versioning the "assemble_batch_operations" method and its dependencies.

Rather:
- unversioned `assemble_batch_operations` method returns the latest version of CommitBatchArguments
- this method checks `api_version()` and determines if it can express the CommitBatchArguments in a compatible version.  In the first case that we have to handle, that means: `v1::CommitBatchArguments` can be trivially converted to `v0::CommitBatchArguments` if there are no operations that set properties.

This will reduce greatly the volume of code duplication and version handling.

Reply:
I like where this is going.

One thing that is unclear to me is what should happen when `v1::CommitBatchArguments` contains `SetAssetProperties`, but the `api_version()` returns `0`? Should we just "force-convert" from v1 to v0 by dropping all of the `SetAssetProperties` from `v1::CommitBatchArguments`?